packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart

```diff
@@ -121,17 +121,38 @@ abstract class RenderSliverBoxChildManager {
   /// true without making any assertions.
   bool debugAssertChildListLocked() => true;
 }
-
-/// Parent data structure used by [RenderSliverMultiBoxAdaptor].
-class SliverMultiBoxAdaptorParentData extends SliverLogicalParentData with ContainerParentDataMixin<RenderBox> {
-  /// The index of this child according to the [RenderSliverBoxChildManager].
-  int index;
+/// Parent data structure used by [RenderSliverWithKeepAliveMixin].
+abstract class KeepAliveParentDataMixin extends ParentData {
+  // This class is intended to be used as a mixin, and should not be
+  // extended directly.
+  factory KeepAliveParentDataMixin._() => null;
 
   /// Whether to keep the child alive even when it is no longer visible.
   bool keepAlive = false;
 
   /// Whether the widget is currently being kept alive, i.e. has [keepAlive] set
   /// to true and is offscreen.
+  bool get keptAlive;
+}
+
+// This class exists to dissociate [KeepAlive] from [RenderSliverMultiBoxAdaptor].
+abstract class RenderSliverWithKeepAliveMixin extends RenderSliver {
+  // This class is intended to be used as a mixin, and should not be
+  // extended directly.
+  factory RenderSliverWithKeepAliveMixin._() => null;
+
+  @override
+  void setupParentData(RenderObject child) {
+    assert(child.parentData is KeepAliveParentDataMixin);
```

Comment:
Does this work?  I think most callers will expect this to create the parent data if it's null.  We can't really do that with a mixin though.

Reply:
The only purpose of this, is to alert the developer that we should use a `KeepAliveParentDataMixin `as a parentData for the child. This method have to be overridden by classes applying the `RenderSliverWithKeepAliveMixin` mixin. But yes, it would be better if we could create the parent data if it's null, but I don't know how to do it with mixins.

Reply:
I started this PR with an abstract class, then @Hixie suggested me to use mixins instead, because the KeepAlive feature is orthogonal to the child model. I think it's make more sense to have a mixin for this, but like you, I'm not satisfied with this code 😞 .

Reply:
Couldn't we just enforce that by making this an abstract class with an abstract method?  If that's too restrictive, what value does this mixin actually add?

for instance, if someone mixing it in calls `super.setupParentData(child)` before they actually create the parent data, they'll get a runtime failure but only in debug mode.

Reply:
The point of using the mixin is that you don't know what the superclass should be.

Reply:
(at the point where you're writing this code)
that's sort of the whole point of this PR, as i understand it. :-)